### ISSUE TITLE
Support for DDS (Microsoft DirectDraw Surface)

### ISSUE THREAD

**Issue by p4paul** [haraldk/TwelveMonkeys#1013]
DDS is a raster image. Whilst it is typically used for textures I want to show previews for the files.

I found a simple MIT implementation, neatly contained in one Java class... 
https://github.com/npedotnet/DDSReader/blob/master/src/java/DDSReader.java

Simple test using the DDSReader class:
```
		String file = "src/test/resources/files/textures/dds/dxt5.dds";
		try (InputStream stream = new FileInputStream(file)) {
			byte[] buffer = new byte[stream.available()];
			stream.read(buffer);

			int[] pixels = DDSReader.read(buffer, DDSReader.ARGB, 0);
			int width = DDSReader.getWidth(buffer);
			int height = DDSReader.getHeight(buffer);
			BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
			image.setRGB(0, 0, width, height, pixels, 0, width);
		}
```

Not sure what effort would be needed to wrap the class in the `imageio` framework, but it would make a nice addition.

**Comment by p4paul** [comment]
Hi Harald,

I'll take a look and see if I can carve out a bit of time.  I was going to base it on PCX Image reader, looks simple enough as a starting point.

**Comment by haraldk** [comment]
Good plan! I think the HDR plugin is also a good plugin to look at, as it's based on delegating most work to the `RGBE` class, much like you would delegate to the `DDSReader`. 

Good luck! And don't hesitate to ask, if there's something I can help with, regarding ImageIO API specific issues.

**Comment by p4paul** [comment]
Is there a way to swap endian when using imageInput.readInt()?

**Comment by p4paul** [comment]
Found it...
imageInput.setByteOrder(ByteOrder.LITTLE_ENDIAN);

**Comment by p4paul** [comment]
Ok have a very nasty implementation... 

<img width="563" alt="Screenshot 2024-09-13 at 15 40 45" src="https://github.com/user-attachments/assets/89d90844-23a4-42fa-86a1-5dc6d129cb8a">

**Comment by haraldk** [comment]
That was quick! 😀

**Comment by haraldk** [comment]
I think the code looks mostly good, but I have a few comments/suggestions... 😉 

In `DDSHeader.read`: 

* Verify Magic!  
* Consider verifying `dwSize == 124` 
* Probably some more, like verifying that we have valid DXTn values etc

Also, if you keep the `mark/reset` code (more on this below), consider wrapping all the reading in a `try/finally` block:

```java
imageInput.mark();
try {
    imageInput.setByteOrder(ByteOrder.LITTLE_ENDIAN);

    byte[] magic = new byte[4];
    imageInput.readFully(magic);

    // ...more header paring...
}
finally {
    imageInput.reset();
}
```

In `DDSImageReader`

`readHeader`:
The `flush/seek` is a bit odd, as the `DDSHeader.read` resets the stream back to start just before this. But makes sense if you plan to move the header reading out of the `DDSReader` class. 😀 

`getRawImageType`:
Override to 

```java
    return ImageTypeSpecifiers.createFromBufferedImageType(BufferedImage.TYPE_INT_ARGB);
```

…which is the pixel format you use in the `read` method. You can then expand on this later, to create a pixel format that is most suitable for the image data.

`getImageTypes`:
Using `float` data type is not a good idea here… This is probably a copy & paste bug...😉 Instead, change to simply (assuming the `getRawImageType` method is changed as suggested above):

```
    return Collections.singletonList(getRawImageType(imageIndex)); 
```

`read`:

With the above changes, `destination` will already be of `TYPE_INT_ARGB` (as we need),  you should no longer create a new `BufferedImage`, but just use `destination` directly, and copy the pixels into it. This is required for the API to work 100% (there should be tests covering it as well, but not sure 😬 ). For other formats than the ARGB format, we need to copy pixels through the raster, but  using `setRGB` is okay for ARGB.

Another important thing, using `imageInput.read(buffer)` is not safe here, and may result in only a partially filled array. The best way to fix this, would be to compute the number of bytes to read, based on the header values, and then use `readFully`. If computing the length of the buffer isn’t feasible, it may be possible to just read all the bytes until the end of the stream using a loop (ie. check the length of bytes read by the `read` method, and invoke it until you get `-1`).

Finally, we should probably add test cases for all the different DXTn formats. 😀

**Comment by haraldk** [comment]
Looks good! 👍🏻 

Regarding the index out of bounds exception, don't worry too much right now. The ImageIO API requires readers to support subsampled reading and/or reading only a specified region of the image. But we can always fake that if needed, by reading the entire image first, and extract the correct pixels later.

**Comment by p4paul** [comment]
What are the next steps?  Would you like me to create a PR or are there any specific issues I should address in my fork?
I would also like to get an idea about possible release windows, assuming there is not too much additional work needed.

**Comment by haraldk** [comment]
If you create a PR, I'll have a look! 

I belive it could be merged and released quite soon (ie. next week) unless there's some major issues. We need to make the tests pass, but we'll figure that out. If there are no changes in code outside the new module, it should be quite straight forward.

**Comment by p4paul** [comment]
I have created PR https://github.com/haraldk/TwelveMonkeys/pull/1021

**Comment by haraldk** [comment]
Excellent! 

I've been completely busy this weekend, but I'll have a look as soon as possible!

**Comment by p4paul** [comment]
Thank you - I was hoping to make use of the DDS feature around November, so sounds like there should be plenty of time.  Do let me know if there is any doc or Legal attribution files you would like me to update (MIT for [DDSReader.java](https://github.com/npedotnet/DDSReader/blob/master/src/java/DDSReader.java))?

**Comment by haraldk** [comment]
We should probably add a `license.txt` in the root folder of the module, containing both the standard TwelveMonkeys BSD license + acknowledge the use of `DDSReader.java` and include the accompanying MIT license. As these licenses are generally comparable I don't think we need to do much more.

**Comment by haraldk** [comment]
Excellent work! I added some missing pieces and re-added all the tests. A SNAPSHOT build should be available right away. A release should be out soon. 

Closed via #1021

**Comment by haraldk** [comment]
3.12.0 with your DDS support is now released. 😀
